qiskit/optimization/problems/optimization_problem.py

```diff
@@ -35,203 +41,188 @@ class OptimizationProblem:
-
-        # set linear objective terms
-        for i, v in self.objective.get_linear().items():
-            op.objective.set_linear(i, v)
-
-        # set quadratic objective terms
-        for i, v_i in self.objective.get_quadratic().items():
-            for j, v in v_i.items():
-                op.objective.set_quadratic_coefficients(i, j, v)
-
-        # set objective offset
+        op.objective.set_sense(self.objective.get_sense())
         op.objective.set_offset(self.objective.get_offset())
+        op.objective.set_linear([(i, v) for i, v in self.objective.get_linear_dict().items()])
+        if self.objective.get_num_quadratic_nonzeros() > 0:
+            op.objective.set_quadratic(self.objective.get_quadratic())
 
         # set linear constraints
-        linear_rows = self.linear_constraints.get_rows()
-        linear_sense = self.linear_constraints.get_senses()
-        linear_rhs = self.linear_constraints.get_rhs()
-        linear_ranges = self.linear_constraints.get_range_values()
-        linear_names = self.linear_constraints.get_names()
-        op.linear_constraints.add(linear_rows, linear_sense, linear_rhs,
-                                  linear_ranges, linear_names)
-
-        # TODO: add quadratic constraints
-
+        lin_expr = self.linear_constraints.get_rows()
+        senses = self.linear_constraints.get_senses()
+        rhs = self.linear_constraints.get_rhs()
+        range_values = self.linear_constraints.get_range_values()
+        names = self.linear_constraints.get_names()
+        op.linear_constraints.add(
+            lin_expr=lin_expr, senses=senses, rhs=rhs, range_values=range_values, names=names)
+
+        # set quadratic constraints
+        names = self.quadratic_constraints.get_names()
+        senses = self.quadratic_constraints.get_senses()
+        rhs = self.quadratic_constraints.get_rhs()
+        lin_expr = self.quadratic_constraints.get_linear_components()
+        quad_expr = self.quadratic_constraints.get_quadratic_components()
+        for i in range(self.quadratic_constraints.get_num()):
+            op.quadratic_constraints.add(
+                lin_expr=lin_expr[i], quad_expr=quad_expr[i], sense=senses[i], rhs=rhs[i],
+                name=names[i])
         return op
 
     def end(self):
         """Releases the OptimizationProblem object."""
-        if self._disposed:
-            return
-        self._disposed = True
-
-    def __del__(self):
-        """non-public"""
-        self.end()
+        self._name = ''
+        self.variables = VariablesInterface()
+        varindex = self.variables.get_indices
+        self.linear_constraints = LinearConstraintInterface(varindex=varindex)
+        self.quadratic_constraints = QuadraticConstraintInterface(varindex=varindex)
+        self.objective = ObjectiveInterface(varindex=varindex)
+        self.solution = SolutionInterface()
+        self._problem_type = None
 
     def __enter__(self):
```

Comment:
Add type hint, note that you need to add an import for this to work, since using the type of the class itself is not yet in the standard python:
```
from __future__ import annotations
```

```suggestion
    def __enter__(self) -> OptimizationProblem:
```

Reply:
ok

Reply:
Yes you're right, the future import works only in 3.7 onwards -- sorry, I forgot! Then using the string is the right way. That's also done in other places in Aqua.

The QASMParser you showed doesn't have type hints because Terra didn't include them in existing files yet. We only started recently to use type hints in Qiskit. Therefore they are not everywhere yet but we're trying to put them in all code that is new.

Reply:
Done